crates/turbo-tasks-fs/src/lib.rs

```diff
@@ -80,6 +80,64 @@ pub trait FileSystem: ValueToString {
     fn metadata(&self, fs_path: FileSystemPathVc) -> FileMetaVc;
 }
 
+#[derive(Default)]
+struct DiskWatcher {
+    watcher: Mutex<Option<RecommendedWatcher>>,
+    /// Keeps track of which directories are currently watched. This is only
+    /// used on a OS that doesn't support recursive watching.
+    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
+    watching: dashmap::DashSet<PathBuf>,
+}
+
+impl DiskWatcher {
+    #[cfg(not(any(target_os = "macos", target_os = "windows")))]
+    fn restore_if_watching(&self, dir_path: &Path, root_path: &Path) -> Result<()> {
+        if self.watching.contains(dir_path) {
+            let mut watcher = self.watcher.lock().unwrap();
```

Comment:
We have a lock that's only used inside a loop. If we passed the `HashSet` to this method, we could lock once and iterate